nisapi/clean.py

```diff
@@ -0,0 +1,291 @@
+            "demographic_level": "demographic_type",
+            "demographic_name": "demographic_value",
+            "indicator_label": "indicator_type",
+            "indicator_category_label": "indicator_value",
+        }
+    )
+
+
+def col_values_in(df: pl.DataFrame, col: str, values: str) -> bool:
+    """All values of `df` in column `col` are in `values`?"""
+    return df[col].is_in(values).all()
+
+
+def remove_near_duplicates(
+    df: pl.DataFrame,
+    tolerance: float,
+    filter_expr: pl.Expr,
+    n_fold_duplication: int = None,
+    value_columns: Sequence[str] = ["estimate", "ci_half_width_95pct"],
+    group_columns: Sequence[str] = None,
+) -> pl.DataFrame:
+    """Remove near-duplicate rows
+
+    Two rows are "near-duplicate" if they have all the same grouping variables
+    (i.e., everything other than estimate and CI) and have estimate and CI
+    values that are within some tolerance of one another.
+
+    Args:
+        df (pl.DataFrame): input data frame
+        tolerance (float): greatest difference in `estimate` and `ci_half_width_95pct`
+        filter_expr (Expr): How to filter for the desired row in each group.
+          Implemented like `dataframe.filter(filter_expr.over(group_columns))`.
+        n_fold_duplication (int, optional): For each set of grouping values,
+          there are exactly this number of near-duplicate rows. If None (default),
+          do not apply this kind of check.
+        value_columns (Sequence[str]): names of the value columns. Defaults to
+          `["estimate", "ci_half_width_95pct"]`.
+        group_columns (Sequence[str]): names of the grouping columns. If None
+          (the default), uses all columns in `df` that are not in `value_columns`.
+
+    Returns:
+        pl.DataFrame: _description_
+    """
+    if group_columns is None:
+        group_columns = set(df.columns) - set(value_columns)
+
+    # a "group" is a unique combination of group columns
+    group_size_col = "group_size"
+    assert group_size_col not in group_columns
+    groups = df.group_by(group_columns).len(name=group_size_col)
+    dup_groups = groups.filter(pl.col(group_size_col) > 1)
+
+    if n_fold_duplication is not None:
+        # all groups that aren't of size 1 should be the "fold" duplication size
+        assert (dup_groups[group_size_col] == n_fold_duplication).all()
+
+    # assert that the estimates and CIs in these groups are similar to
+    # one another, within some small margin
+    assert (
+        dup_groups.drop(group_size_col)
+        .join(df, how="inner", validate="1:m", on=group_columns)
+        .group_by(group_columns)
+        .agg((pl.col(value_columns).pipe(lambda x: x.max() - x.min())))
+        .select((pl.col(value_columns) < tolerance).all())
+        .select(pl.all_horizontal(value_columns))
+        .item()
+    )
+
+    # aggregate to drop the nearly-duplicate rows
+    out = df.filter(filter_expr.over(group_columns))
```

Comment:
I don't see where `filter_expr` is assigned a value, either inside this function (lines 195-220), by default in line 168, or where this function is called in line 103. So I don't understand how this line works to preserve exactly one copy of each near-duplicated row.